### PR TITLE
Implement LSP migration find usages (#396)

### LINKED ISSUE #396: [lsp] Change find usages to use LSP
This is currently implemented with `DartServerFindUsagesHandler`

### PR DESCRIPTION
**Pull request by ranbeuer**
## 📌 Summary

  This PR migrates Find Usages / References from the legacy Dart Analysis Server (DAS) custom search protocol to the JetBrains LSP framework (textDocument/references), as part of the broader LSP
  migration effort.

  When Enable Experimental LSP features is checked under Settings | Languages & Frameworks | Dart, Find Usages queries (textDocument/references) are routed through the in-process DartBridgeLspServer
  and resolved using the JetBrains platform LSP client. When disabled, it falls back to the legacy DartServerFindUsagesHandler.

  ### Relevant Issues

  • Fixes #396
  • Part of #207
  ──────
  ## 🛠️ Changes

- Bridge Capabilities & Forwarding (`DartBridgeLspServer.kt`):
   - Declared `setReferencesProvider(true)` in `ServerCapabilities` during initialize.
   - Implemented `references(ReferenceParams)` to forward `"textDocument/references"` requests to DAS via lsp.handle and parse the returned `List<Location>`.
- Descriptor Customizer (`DartLspServerDescriptor.kt`):
   - Configured `findReferencesCustomizer` to use `LspFindReferencesSupport()` when `DartConfigurable.isExperimentalLspFeaturesEnabled(project)` is enabled.
- Legacy Handler Gating (`DartServerFindUsagesHandlerFactory.java`):
    - Gated `canFindUsages()` to return false when experimental LSP is active so the JetBrains LSP client takes precedence.
- LSP Method __Enum__ (`LspMethod.kt`):
    - Registered `REFERENCES("textDocument/references", isExperimental = true, presentableName = "references")`.
- Unit Tests (`DartBridgeLspServerTest.kt`):
    - Added `testReferencesRequest` verifying end-to-end request forwarding and response deserialization through the mock bridge server.

## 🧪 Manual Testing Instructions
<details>
<summary>Click for manual testing instructions. </summary>

### 1. Launch the Sandbox IDE

  From the root of the repository, execute:
    `./gradlew runIde`

  (Optionally run `./gradlew clean runIde` for a completely fresh sandbox environment).

### 2. Configure Dart SDK and Settings

  1. In the sandbox IDE, open an existing Dart project or create a new Dart command-line application.
  2. Open Settings / Preferences (Ctrl+Alt+S on Linux/Windows, Cmd+, on macOS).
  3. Navigate to Languages & Frameworks → Dart:
      • Ensure Enable Dart support for the project is checked and a valid Dart SDK path is configured.
      • Locate the Enable Experimental LSP features checkbox.
  ### 3. Test Find Usages with LSP (Experimental LSP = ON)

  1. Ensure "Enable Experimental LSP features" is checked and click Apply.
  2. Open a Dart file with multiple symbol usages across files (e.g., a function, class, or variable referenced in bin/ and lib/).
  3. Place the caret on a symbol and invoke Find Usages:
      • Find Usages: Alt + F7 (or Option + F7 on macOS)
      • Show Usages: Ctrl + Alt + F7 (or Cmd + Option + F7 on macOS)
  4. Expected Behavior:
      • The Find Usages tool window (or Show Usages popup) appears.
      • The search target presentation displays the LSP target format (e.g. LSP (filename:line:col)).
      • All references across the project files are correctly listed.
      • Double-clicking any usage navigates directly to the exact file and text range.
  5. Edge Case Verification:
      • Test on external dependencies (e.g. symbols inside .pub-cache or Dart SDK core libraries like dart:io).
      • Test on newly opened files after IDE startup.
  ### 4. Test Fallback to Legacy DAS (Experimental LSP = OFF)

  1. Open Settings → Languages & Frameworks → Dart.
  2. Uncheck "Enable Experimental LSP features" and click Apply.
  3. Place the caret on the same symbol and trigger Find Usages (Alt + F7).
  4. Expected Behavior:
      • Control seamlessly falls back to legacy DartServerFindUsagesHandler.
      • The Find Usages tool window displays results using the legacy grouping and presentation.
  ### 5. Optional: Inspect LSP Traffic in idea.log

  To verify that textDocument/references requests are sent over the bridge:

  1. Go to Help → Diagnostic Tools → Debug Log Settings....
  2. Add #com.intellij.platform.dartlsp.
  3. Trigger Find Usages and check idea.log (Help → Show Log in Files / Finder) to observe the textDocument/references JSON-RPC messages and responses.

</details>

## Screenshots for this change:
298eis12a7b5g (1)

---

Review the contribution guidelines below:

- [X] I’ve reviewed the contributor guide and applied the relevant portions to this PR.
- [X] I've included the required information in the description above.
- [X] I've updated `CHANGELOG.md` if appropriate (i.e. user-facing change)

<details>
  <summary>Contribution guidelines:</summary><br>

- See the [Flutter organization contributor guide](https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md) for general expectations for PRs.
- Larger or significant changes should be discussed in an issue before creating a PR.
- Dart contributions to our repos should follow the [Dart style guide](https://dart.dev/guides/language/effective-dart) and use
  `dart format`.
- Java and Kotlin contributions should strive to follow Java and Kotlin best
  practices ([discussion](https://github.com/flutter/flutter-intellij/issues/8098)).

</details>

### CODE REVIEW

File: third_party/src/test/java/com/jetbrains/lang/dart/lsp/DartBridgeLspServerTest.kt [hunk 1]
```diff
@@ -307,6 +309,52 @@ class DartBridgeLspServerTest : DartCodeInsightFixtureTestCase() {
...
+    fun testReferencesRequest() {
+        val params = ReferenceParams().apply {
+            textDocument = TextDocumentIdentifier("file://test.dart")
+            position = Position(1, 2)
+            context = ReferenceContext(true)
+        }
+        val future = bridgeServer.references(params)
+        val jsonObject = capturedRequests.find { it.get("method")?.asString == "lsp.handle" }
+        assertNotNull("An lsp.handle request should be sent to DAS", jsonObject)
+        assertEquals("123", jsonObject!!.get("id").asString)
```
[the remark this answers is not in the document]
Comment (ranbeuer) [reply third_party/src/test/java/com/jetbrains/lang/dart/lsp/DartBridgeLspServerTest.kt]: Done

File: third_party/src/main/java/com/jetbrains/lang/dart/lsp/DartBridgeLspServer.kt [hunk 2]
```diff
@@ -253,6 +255,12 @@ class DartBridgeLspServer(private val project: Project) : DartLanguageServer, Te
         return forwardRequest<List<DocumentHighlight>>("textDocument/documentHighlight", params, type)
     }
 
+    override fun references(params: ReferenceParams?): CompletableFuture<List<Location>> {
```
[the remark this answers is not in the document]
Comment (ranbeuer) [reply third_party/src/main/java/com/jetbrains/lang/dart/lsp/DartBridgeLspServer.kt]: Done

**Comment by helin24** [comment]
The references LSP-over-legacy endpoint was only added with a recent SDK change, so this will require gating on the Dart SDK version. Can you test with an older Dart SDK as well and make sure that works well?

**Comment by ranbeuer** [comment]
> The references LSP-over-legacy endpoint was only added with a recent SDK change, so this will require gating on the Dart SDK version. Can you test with an older Dart SDK as well and make sure that works well?

Hi @helin24, 

I've verified the feature across a matrix of Dart SDK versions representing all major milestones:

- **Dart 2.14.0** (legacy Dart 2)
- **Dart 2.19.0** (final Dart 2 milestone)
- **Dart 3.0.0** (initial Dart 3 release)
- **Dart 3.12.2** (latest release installed)

### Verification Summary:
- **Default (Experimental LSP Disabled):** Legacy Find Usages continues to work seamlessly across all tested versions.
- **Experimental LSP Enabled:** LSP references (`textDocument/references`) function as expected when enabled.

Everything ran smoothly without any regressions across all four versions.

**Comment by helin24** [comment]
Is this falling back to legacy handling in the earlier Dart versions though? e.g. are there any `MethodNotFound` errors from analysis server? Since the endpoint being used was only added in https://github.com/dart-lang/sdk/commit/c6d728bfd33db2c8832a345b77403998cc627e2b, I want to make sure we understand what's happening in the case of the Dart SDK versions without that change.

**Comment by ranbeuer** [comment]
> Is this falling back to legacy handling in the earlier Dart versions though? e.g. are there any `MethodNotFound` errors from analysis server? Since the endpoint being used was only added in [dart-lang/sdk@c6d728b](https://github.com/dart-lang/sdk/commit/c6d728bfd33db2c8832a345b77403998cc627e2b), I want to make sure we understand what's happening in the case of the Dart SDK versions without that change.

Regarding the question on whether `MethodNotFound` or server exceptions weren't observed on earlier Dart versions:

1. **`lsp.handle` Wrapping**: Requests are not sent directly as raw LSP calls; instead, they are tunneled through the legacy DAS domain command `lsp.handle` (`{ "method": "lsp.handle", "params": {
  "lspMessage": ... } }`). Because `lsp.handle` has been part of DAS since early 3.x, the outer server process always recognizes and accepts the command.
2. **Existing Internal LSP Handler**: Inside DAS, `textDocument/references` was already a recognized method inherited from the standalone LSP implementation. Rather than rejecting the request with an error, DAS processes it and returns standard JSON-RPC responses (or empty/null results), which `DartBridgeLspServer` unwraps gracefully without throwing unhandled exceptions.
3. **Why `3.14.0-65.0.dev` is the Gate**: Even though earlier servers don't crash on the method, `3.14.0-65.0.dev` is the milestone where server-side LSP references were stabilized and verified for full parity with IntelliJ's project model. Version gating ensures we only activate the LSP path when that stability threshold is met, while earlier SDKs continue to rely on the battle-tested legacy DAS handler.


For this, I pushed an update to ensure LSP Find References is properly version-gated and gracefully falls back to legacy DAS on older Dart SDKs:

1. **Minimum SDK Constraint**:
   - Defined `MIN_LSP_REFERENCES_SDK_VERSION = "3.14.0-65.0.dev"` and added `isLspReferencesEnabled(Project)` in `DartAnalysisServerService`.

2. **LSP Descriptor & Legacy Handler Gating**:
   - Updated `DartLspServerDescriptor.findReferencesCustomizer` to check `isLspReferencesEnabled(project)` instead of just the experimental feature flag, ensuring `LspFindReferencesSupport` is only registered when the active SDK supports it.
   - Updated `DartServerFindUsagesHandlerFactory` and `DartFindUsagesProvider` to yield only when `isLspReferencesEnabled(project)` is true, allowing legacy DAS Find Usages to seamlessly take over on Dart < 3.14 without requiring IDE restarts.

3. **Tests & Verification**:
   - Added `testIsDartSdkVersionSufficientForLspReferences()` in `DartBridgeLspServerTest.kt` to verify version constraint checks.
   - Verified live dynamic SDK switching between Dart 3.14 (LSP active) and Dart 3.13 (legacy DAS fallback) in the sandbox IDE.
   - All Find Usages and LSP bridge tests pass cleanly.

File: third_party/CHANGELOG.md [hunk 3]
```diff
@@ -1,4 +1,5 @@
 ## Unreleased
+- Find Usages / references with JetBrains LSP (experimental feature) (#396).
```
Comment (helin24) [inline third_party/CHANGELOG.md]: Move this to one of the subcategories - probably added?

**Review by helin24 — APPROVED** [review]
Thanks for the notes on versioning. This LGTM!

@pq can review and merge after changelog is updated.

**Review by pq — APPROVED** [review]

**Comment by helin24** [comment]
@ranbeuer In the screenshot, it looks like there is a disambiguation request to the user that's unnecessary since both items in the list are the same thing (the first line refers to the line/column of `calculate` and the second line also refers to `calculate`). It seems this is changed from the legacy implementation, which should pull up a find usages tool window right away, e.g.:

image

Is the first or second line from the disambiguation pop up unnecessary in LSP?

**Comment by ranbeuer** [comment]
> @ranbeuer In the screenshot, it looks like there is a disambiguation request to the user that's unnecessary since both items in the list are the same thing (the first line refers to the line/column of `calculate` and the second line also refers to `calculate`). It seems this is changed from the legacy implementation, which should pull up a find usages tool window right away, e.g.:
> 
> image
> Is the first or second line from the disambiguation pop up unnecessary in LSP?

@helin24 To answer your question:
• The 1st line (file:line:col) is the LSP target (`LspSearchTarget`) — this is the necessary one that triggers the LSP `"textDocument/references"` request.
• The 2nd line (calculate()) is a legacy PSI target (`PsiTargetVariant`) — this is the unnecessary one. (In fact, selecting the 2nd line silently does nothing because `canFindUsages` returns false when LSP is enabled). 

Why both appeared:
When `DartFindUsagesProvider.canFindUsagesFor` and `DartServerFindUsagesHandlerFactory.canFindUsages` return `false`, IntelliJ's `SearchTargetVariantsDataRuleKt` sees that `USAGE_TARGETS_KEY` is null and falls back to `TargetElementUtil.findReference(...)`.
Because JetBrains designed platform LSP assuming plain-text/TextMate files without a PSI tree, they didn't expect `findReference` to return anything. However, since Dart files still have a full PSI tree, `TargetElementUtil` resolves the `DartReference` and appends the redundant `PsiTargetVariant` alongside the `LspSearchTarget`, forcing the disambiguation menu. 
Your fix in #643 https://github.com/flutter/dart-intellij-third-party/pull/643 (`DartTargetElementEvaluator.getTargetCandidates` returning `Collections.emptyList()` when LSP is enabled) suppresses that redundant PSI target so IntelliJ only sees the `LspSearchTarget` and opens the usages popup/window immediately. Tested locally and confirmed it works great!
